qcoro/qcorotask.h

```diff
@@ -742,23 +772,44 @@ inline T waitFor(QCoro::Task<T> &&task) {
  *        For void tasks, it needs to have no arguments.
  *        For all other types, it takes a value of the type as single argument.
  */
-template <typename T, typename Callback>
-requires std::is_invocable_v<Callback> || std::is_invocable_v<Callback, T>
-void connect(QCoro::Task<T> &task, QObject *context, Callback func) {
+template <typename T, typename QObjectSubclass, typename Callback>
+requires std::is_invocable_v<Callback> || std::is_invocable_v<Callback, T> || std::is_invocable_v<Callback, QObjectSubclass>
+void connect(QCoro::Task<T> &&task, QObjectSubclass *context, Callback func) {
     QPointer ctxWatcher = context;
     if constexpr (std::is_same_v<T, void>) {
-        task.then([ctxWatcher, func = std::move(func)]() {
+        task.then([ctxWatcher, func = std::move(func), context]() {
             if (ctxWatcher) {
-                func();
+                if constexpr (std::is_member_function_pointer_v<Callback>) {
+                    (context->*func)();
+                } else {
+                    Q_UNUSED(context); // clang checks for unused variables after computing the constexpr if,
+                                       // which leads to a false-positive here
+                    func();
+                }
             }
         });
     } else {
-        task.then([ctxWatcher, func = std::move(func)](auto &&value) {
+        task.then([ctxWatcher, func = std::move(func), context](auto &&value) {
             if (ctxWatcher) {
-                func(std::forward<decltype(value)>(value));
+                if constexpr (std::is_member_function_pointer_v<Callback>) {
+                    (context->*func)(std::forward<decltype(value)>(value));
+                } else {
+                    Q_UNUSED(context); // clang checks for unused variables after computing the constexpr if,
+                                       // which leads to a false-positive here
+                    func(std::forward<decltype(value)>(value));
+                }
```

Comment:
This has actually been broken before, but it looks like if the `func` takes no arguments (that is, `std::is_invocable_v<Callback>` is satisfied), this code will fail.

I think this case needs an additional if branch  for each case which will silently discard the `value` if the `Callback` does not take any arguments.

Since you are touching this code, would you mind fixing it?